File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int[][] answer = new int[triangle.length][triangle[triangle.length-1].length];
        
        answer[0][0] = triangle[0][0];
        
        for(int row = 1; row< triangle.length; row++){
            for(int col = 0; col< triangle[row].length; col++){
                // 가장 좌측
                if(col==0)
                    answer[row][0] = answer[row-1][0] + triangle[row][0];
                // 가장 우측
                else if(col == triangle[row].length){
                    answer[row][col] = answer[row-1][col-1] + triangle[row][col];
                }
                // 중간
                else{
                    answer[row][col] 
                        = Math.max(
                        answer[row-1][col-1] + triangle[row][col],
                        answer[row-1][col] + triangle[row][col]
                    );
                }
                
            }
        }
        int result = Integer.MIN_VALUE;
        for(int res: answer[triangle.length-1])
            result = Math.max(result,res);
        return result;
    }
}

    

File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    dp = [[0 for i in range(len(triangle))]for i in range(len(triangle))]
    dp[0][0] = triangle[0][0]
    
    for row in range(1,len(triangle)):
        for col in range(len(triangle[row])):    
            if col == 0:
                dp[row][col] = triangle[row][col] + dp[row-1][0]
            elif col == len(triangle[row])-1:
                dp[row][col] = triangle[row][col] + dp[row-1][col-1]
            else:
                dp[row][col] = max(dp[row-1][col]+triangle[row][col],dp[row-1][col-1]+triangle[row][col])

    return max(dp[len(triangle[-1])-1])